witness.md

```diff
@@ -0,0 +1,599 @@
+type Witness = (Node|Instruction...)
+```
+
+
+## Execution Enviroment
+
+The witness execution environment MUST contain the following 2 elements:
+
+- **WitnessHeader** -- a header containing the version of the witness. The `version` MUST be 1.
+
+- **Witness** -- a witness to be executed;
+
+- **Substitution Rules** -- a list of all possible substitution rules.
+
+
+## Execution process
+
+Initially, the Witness MUST BE an array of `Instruction`s.
+
+Then, as substitution rules are applied to the witness, some elements of the
+array are replaces with `Node`s.
+
+The execution continues until there are no substitution rules left to execute.
+
+Here is how the execution code might look like in Go for building a single trie.
+
+```go
+witness := GetInitialWitness()
+rules := GetSubstitutionRules()
+numberOfRulesApplied := 1 // initial state
+
+for numberOfRulesApplied > 0 {
+    witness, numberOfRulesApplied := ApplyRules(witness, rules)
+}
+
+if len(witness) == 1 {
+    trie.root = witness[0]
+} else {
+    panic("witness execution failed")
+}
+
+```
+
+
+And here is an example of the execution process (we will use the set of rules
+form the **Substitution Rules** section of this document):
+
+* **Step 1**. Witness: `(HASH h1 HASH h2 BRANCH 0b101 HASH h3 BRANCH 0b11)`
+
+* **Step 2**. Apply `HASH` substitution rules.
+Witness: `(HashNode{h1} HashNode{h2} BRANCH 0b101 HashNode{h3} BRANCH 0b11)`
+
+* **Step 3**. Apply `BRANCH` substitution rules (only once, because `BRANCH 0b11`
+doesn't pass its `GUARD` statements just yet).
+Witness: `(BranchNode{0: HashNode{h1} 2:HashNode{h2}} HashNode{h3} BRANCH 0b11)`
+
+* **Step 4**. Apply `BRANCH` substitution rules again.
+Witness: `(BranchNode{0: BranchNode{0: HashNode{h1} 2:HashNode{h2}} 1:HashNode{h3}})`
+
+* **Step 5**. No more rules are applicable, the witness contains only one
+    element, the execution ends successfully.
+
+
+## End Criteria
+
+The execution ends when there are no substitution rules applicable for this
+witness.
+
+### Building a single trie from the witness
+
+If we are building a single trie from the witness, then the only SUCCESS
+executon is when the following are true:
+
+- The execution state MUST match the End Criteria
+- There MUST be only one item left in the witness
+- This item MUST be of the type `Node`
+    
+In that case, this last item will be the root of the built trie.
+
+Every other end state is considered a FAILURE.
+
+
+### Building a Forest 
+
+We also can build a forest of tries with this approach, by adding a new
+Instruction `NEW_TRIE` and adjusting the success criteria a bit:
+
+- The execution state MUST match the End Criteria;
+- The items that are left in the witness MUST follow this pattern: `(Node
+    NEW_TRIE ... Node)`
+- Each `Node` element will be a root of a trie.
+
+Every other end state is considered a FAILURE.
+
+
+## Instructions & Parameters
+
+A single instruction consists of substitution rules and parameters.
```

Comment:
yeah, I see that not keeping the boundaries between the witness format itself and the `Witness -> Trie` parsing algorithm leaves a lot of ambiguity. I think it will be worth it to split it.